My fruit shop

petro2061 opened the pull request:
Iimplemented logic for creating a fruit shop performance report. The input data is a cdv file containing information on all transactions with a certain type of fruit. The output file contains information about the remaining fruit after all transactions have been completed.
SOLID principles are used for implementation. A model containing information about the fruit and transactions with it was created based on data from the file FruitTransaction class. To call a certain type of handler representing a certain transaction, such a pattern as a strategy implemented on the basis of a map, containing the key - the type of operation, the value of the handler object, is used.
Auxiliary classes for reading and writing data to a file are also implemented. That temporary database FruitStorage class

Vasul-Pankiv commented on src/main/java/core/basesyntax/Main.java:
```diff
@@ -0,0 +1,49 @@
+
+public class Main {
+    public static void main(String[] args) {
+        Map<FruitTransaction.Operation, OperationHandler> strategy = new HashMap<>();
+        strategy.put(FruitTransaction.Operation.BALANCE, new BalanceOperationHandler());
+        strategy.put(FruitTransaction.Operation.SUPPLY, new SupplyOperationHandler());
+        strategy.put(FruitTransaction.Operation.PURCHASE, new PurchaseOperationHandler());
+        strategy.put(FruitTransaction.Operation.RETURN, new ReturnOperationHandler());
```
use Map.of()

Vasul-Pankiv commented on src/main/java/core/basesyntax/impl/FruitTransactionParserImpl.java:
```diff
@@ -0,0 +1,54 @@
+                .map(s -> s.split(SPLIT_DELIMITER))
+                .toList();
+
+        List<FruitTransaction> convertList = new ArrayList<>();
```
```suggestion
        List<FruitTransaction> transactionList= new ArrayList<>();
```

Vasul-Pankiv commented on src/main/java/core/basesyntax/strategy/OperationHandler.java:
```diff
@@ -0,0 +1,8 @@
+import core.basesyntax.model.FruitTransaction;
+
+public interface OperationHandler {
+    void fruitTransactionHandler(FruitTransaction transaction);
```
```suggestion
    void handleOperation(FruitTransaction transaction);
```

Vasul-Pankiv commented on src/main/java/core/basesyntax/servise/ReaderService.java:
```diff
@@ -0,0 +1,7 @@
+import java.util.List;
+
+public interface ReaderService {
+    List<String> readFromFile(String filePath);
```
```suggestion
    List<String> read(String filePath);
```

Vasul-Pankiv commented on src/main/java/core/basesyntax/servise/ReaderService.java:
spell mistake in service package name

Vasul-Pankiv replied:
also move impl package inside service package

Vasul-Pankiv commented on src/main/java/core/basesyntax/impl/ReaderServiceImplFromCsv.java:
```diff
@@ -0,0 +1,26 @@
+import java.nio.file.Path;
+import java.util.List;
+
+public class ReaderServiceImplFromCsv implements ReaderService {
```
```suggestion
public class CsvFileReader implements ReaderService {
```

Vasul-Pankiv commented on src/main/java/core/basesyntax/impl/WriterServiceImplToCsv.java:
```diff
@@ -0,0 +1,19 @@
+import java.io.IOException;
+import java.nio.file.Path;
+
+public class WriterServiceImplToCsv implements WriterService {
```
change naming here too

Vasul-Pankiv reviewed (commented):
pay attention to naming, it always has to describe what exactly method/class/interface is resposible for

Vasul-Pankiv reviewed (changes requested):
check all your namings

petro2061 commented:
Сorrected method and class names according to mentor's comments

okuzan commented on src/main/java/core/basesyntax/Main.java:
```diff
@@ -0,0 +1,48 @@
+        FileReader readerFromFile = new CsvFileReader();
+
+        List<String> dataFromFile = readerFromFile
+                .read("src\\main\\resources\\inputFile.csv");
```
better extract to constant

okuzan commented on src/main/java/core/basesyntax/Main.java:
```diff
@@ -0,0 +1,48 @@
+        String string = report.reportCreator(FruitStorage.fruitTransactionStorage);
+
+        FileWriter writerService = new CsvFileWriter();
+        writerService.write(string, "src\\main\\resources\\outputFile.csv");
```
same

okuzan commented on src/main/java/core/basesyntax/db/FruitStorage.java:
```diff
@@ -0,0 +1,21 @@
+                .orElseThrow(() -> new RuntimeException("Can not find quantity of fruit :"
+                        + fruit
+                        + " in database"));
+
```
```suggestion
```

okuzan commented on src/main/java/core/basesyntax/db/FruitStorage.java:
```diff
@@ -0,0 +1,21 @@
+
+    public static int getQuantity(String fruit) {
+        return fruitTransactionStorage.entrySet().stream()
+                .filter(k -> k.getKey().equals(fruit))
```
avoid short one-letter naming in stream chains outside of math context
change to `fruitKey` or `fruitName`

okuzan commented on src/main/java/core/basesyntax/db/FruitStorage.java:
```diff
@@ -0,0 +1,21 @@
+    public static int getQuantity(String fruit) {
+        return fruitTransactionStorage.entrySet().stream()
+                .filter(k -> k.getKey().equals(fruit))
+                .flatMapToInt(v ->
+                        IntStream.of(v.getValue()))
```
v to quantity

okuzan commented on src/main/java/core/basesyntax/db/FruitStorage.java:
```diff
@@ -0,0 +1,21 @@
+                .flatMapToInt(v ->
+                        IntStream.of(v.getValue()))
+                .findFirst()
+                .orElseThrow(() -> new RuntimeException("Can not find quantity of fruit :"
```
let's throw more specific exceptions where possible, here - NoSuchElementException

okuzan commented on src/main/java/core/basesyntax/model/FruitTransaction.java:
```diff
@@ -0,0 +1,88 @@
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) {
+            return true;
+        }
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+        FruitTransaction that = (FruitTransaction) o;
+        return quantity == that.quantity
+                && Objects.equals(fruit, that.fruit)
+                && operation == that.operation;
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(fruit, quantity, operation);
+    }
```
do you need them?

okuzan commented on src/main/java/core/basesyntax/service/impl/CsvFileWriter.java:
```diff
@@ -0,0 +1,18 @@
+    @Override
+    public void write(String report, String filePath) {
+        Path path = Path.of(filePath);
+        try (java.io.FileWriter fileWriter = new java.io.FileWriter(filePath)) {
```
import instead of writing fully qualified name

okuzan commented on src/main/java/core/basesyntax/service/impl/FruitTransactionParserImpl.java:
```diff
@@ -0,0 +1,54 @@
+    @Override
+    public List<FruitTransaction> converterToTransactions(List<String> readData) {
+        List<String[]> collect = readData.stream()
+                .skip(1)
+                .map(s -> s.split(SPLIT_DELIMITER))
```
```suggestion
                .skip(OFFSET)
                .map(line -> line.split(SPLIT_DELIMITER))
```

okuzan commented on src/main/java/core/basesyntax/service/impl/FruitTransactionParserImpl.java:
```diff
@@ -0,0 +1,54 @@
+        for (String[] strings : collect) {
+            String fruitName = strings[INDEX_FOR_NAME];
+            int quantity = Integer.parseInt(strings[INDEX_FOR_QUANTITY]);
+            String operationType = strings[INDEX_FOR_OPERATION]
+                    .substring(strings[INDEX_FOR_OPERATION].length() - 1);
```
```suggestion
            String operationType = strings[INDEX_FOR_OPERATION];
```
not sure what you need it for

okuzan commented on src/main/java/core/basesyntax/service/impl/FruitTransactionParserImpl.java:
```diff
@@ -0,0 +1,54 @@
+                    + "\"");
+        };
+    }
+
```
```suggestion
```

okuzan commented on src/main/java/core/basesyntax/service/impl/FruitTransactionParserImpl.java:
```diff
@@ -0,0 +1,54 @@
+        return transactionList;
+    }
+
+    private FruitTransaction.Operation getOperationType(String operationType) {
+        return switch (operationType) {
+            case "b" -> FruitTransaction.Operation.BALANCE;
+            case "r" -> FruitTransaction.Operation.RETURN;
+            case "p" -> FruitTransaction.Operation.PURCHASE;
+            case "s" -> FruitTransaction.Operation.SUPPLY;
```
Instead of this, you need to create a method inside of enum, for-each iterate through enum values and compare, throw IllegalArgument if nothing is found. Here you are hardcoding enum codes, which is bad, you make this method tightly coupled to enum - if the enum was to be modified, such as by adding a new operation or changing the code for an operation, the getOperationType method would also need to be updated to reflect these changes. This creates a risk of errors if changes are made in one place and not the other.

By moving the logic of converting a string to an enum instance into the enum itself, through a method that iterates over all enum values and finds the corresponding one, you decouple the getOperationType method from the specific enum codes. The enum becomes responsible for knowing its own string representations, which encapsulates this knowledge within the enum itself.

If the enum provides a method like fromCode(String code), any part of the code that needs to convert a string to the enum can use this method, and if changes are necessary, they're only done within the enum. This follows the principle of single responsibility and makes the system more maintainable.

okuzan commented on src/main/java/core/basesyntax/service/impl/FruitTransactionProcessorImpl.java:
```diff
@@ -0,0 +1,25 @@
+
+public class FruitTransactionProcessorImpl implements TransactionProcessor {
+    @Override
+    public void process(List<FruitTransaction> fruitTransactionsList,
```
```suggestion
    public void process(List<FruitTransaction> fruitTransactions,
```
imo we usually don't include type in variable's name, java is statically typed anyway

okuzan commented on src/main/java/core/basesyntax/service/impl/ReportCreatorImpl.java:
```diff
@@ -0,0 +1,22 @@
+import java.util.Map;
+
+public class ReportCreatorImpl implements ReportCreator {
+    private static final String TITLE_ROW = "\t" + "fruit,quantity" + "\n";
```
system.lineSeparator here too

okuzan commented on src/main/java/core/basesyntax/strategy/BalanceOperationHandler.java:
```diff
@@ -0,0 +1,14 @@
+import core.basesyntax.model.FruitTransaction;
+
+public class BalanceOperationHandler implements OperationHandler {
+
```
```suggestion
```
no empty lines after method/class declaration, per Mate checkstyle rules

okuzan commented on src/main/java/core/basesyntax/strategy/BalanceOperationHandler.java:
```diff
@@ -0,0 +1,14 @@
+
+    @Override
+    public void handlerOperation(FruitTransaction transaction) {
+
```
```suggestion
```

okuzan commented on src/main/java/core/basesyntax/strategy/OperationHandler.java:
```diff
@@ -0,0 +1,8 @@
+
+public interface OperationHandler {
+    void handlerOperation(FruitTransaction transaction);
+
```
```suggestion
```

okuzan commented on src/main/java/core/basesyntax/strategy/PurchaseOperationHandler.java:
```diff
@@ -0,0 +1,21 @@
+    @Override
+    public void handlerOperation(FruitTransaction transaction) {
+        int quantityInTransaction = transaction.getQuantity();
+        if (quantityInTransaction <= FruitStorage.getQuantity(transaction.getFruit())) {
+            FruitStorage.fruitTransactionStorage.put(transaction.getFruit(),
+                    FruitStorage.getQuantity(transaction.getFruit())
+                            - quantityInTransaction);
+        } else {
+            throw new RuntimeException("Not enough quantity for sell: fruit - "
+                    + transaction.getFruit()
+                    + " | quantity - "
+                    + transaction.getQuantity());
+        }
```
in such cases, throwing/returning early and unnesting the rest of your code is [a preferred approach ](https://softwareengineering.stackexchange.com/questions/18454/should-i-return-from-a-function-early-or-use-an-if-statement) (because you don't need `else` after throwing/returning)

okuzan commented on src/main/java/core/basesyntax/service/impl/ReportCreatorImpl.java:
```diff
@@ -0,0 +1,22 @@
+
+    @Override
+    public String reportCreator(Map<String, Integer> inputData) {
+        StringBuilder builderReportRow = new StringBuilder(TITLE_ROW);
```
```suggestion
        StringBuilder reportBuilder = new StringBuilder(TITLE_ROW);
```

okuzan reviewed (changes requested):
Good job! Code style fixes suggested

petro2061 commented:
Fixed problem with naming variables and text style. Also added new method fromCode(String code) which returns enam depending on the type of transaction.
